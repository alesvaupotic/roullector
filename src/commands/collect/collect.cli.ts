import {
  Command,
  Option,
} from 'commander';

import { collect } from '$commands/collect/collect';
import {
  defaultCliCollectOptions,
  keyTransformCliToFunc,
} from '$commands/collect/collect.constants';
import type {
  CollectOptions,
  CliCollectOptions,
} from '$commands/collect/collect.types';

export function collectCli() {
  return new Command('collect')
    .description('collect route data')
    .option(
      '-i, --inDir <path>',
      'directory path to collect',
      defaultCliCollectOptions.inDir
    )
    .option(
      '-e, --extensions <list>',
      'comma-separated list of file extensions to collect',
      defaultCliCollectOptions.extensions,
    )
    .option(
      '-x, --ignorePatterns <list>',
      'comma-separated list of filename patterns (string literal or regex) to ignore',
      defaultCliCollectOptions.ignorePatterns,
    )
    .option(
      '--no-output',
      'don\'t write to disk but print to console instead',
      defaultCliCollectOptions.output,
    )
    .option(
      '-o, --outDir <path>',
      'outputs generated files to this folder',
      defaultCliCollectOptions.outDir,
    )
    .option(
      '-d, --depth <number>',
      'depth of inDir to collect',
      defaultCliCollectOptions.depth,
    )
    .addOption(
      new Option( '-t, --keyTransform <variant>', 'how to transform route key')
        .choices(['none', 'camelCase'])
        .default(defaultCliCollectOptions.keyTransform)
    )
    .option(
      '-k, --dirkey <name>',
      'how to transform route key',
      defaultCliCollectOptions.dirkey,
    )
    .option(
      '--no-utils',
      'don\'t generate utils for building path with arguments',
      defaultCliCollectOptions.utils,
    )
    .option(
      '--no-typescript',
      'outputs files in javascript instead of typescript',
      defaultCliCollectOptions.typescript,
    )
    .option(
      '--verbose',
      'prints more info during operation',
      defaultCliCollectOptions.verbose,
    )
    .action((rawOptions: CliCollectOptions) => {
      const options: CollectOptions = {
        ...rawOptions,
        extensions: rawOptions.extensions?.split(',') ?? [],
        ignorePatterns: rawOptions.ignorePatterns?.split(',') ?? [],
        depth: Number(rawOptions.depth) || Infinity,
        keyTransform: keyTransformCliToFunc[rawOptions.keyTransform],
      };
      if (options.verbose) {
        console.info('Running "collect" with options:');
        console.table(Object.fromEntries(
          Object.entries(rawOptions).map(
            ([key, value]) => ([
              key,
              { value },
            ])
          )),
        );
      }

      const output = collect(options);
      if (!options.output) {
        console.log(output);
      }
    });
}
