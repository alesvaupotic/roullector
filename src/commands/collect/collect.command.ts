import {
  statSync,
  readdirSync,
  writeFileSync,
  mkdirSync,
  existsSync,
  rmSync,
} from 'fs';
import {
  basename,
  join,
  resolve,
} from 'path';

import type { CollectOptions } from './collect.types';

export type DeepObject<T> = {
  [key: string]: T | DeepObject<T>;
};

function extractRouteMapping(options: CollectOptions, inputDir: string = options.inDir, currentPath = '', level = 0): DeepObject<string> {
  const children = readdirSync(inputDir);

  const map: DeepObject<string> = {};

  if (level <= options.depth) {
    for (const child of children) {
      const absPath = join(inputDir, child);
      const stats = statSync(absPath);

      // skip if any ignore pattern matches
      const filename = basename(child);
      if (options.ignorePatterns.some(pattern => new RegExp(pattern).test(filename))) continue;

      if (stats.isDirectory()) {
        const nextPath = `${currentPath}/${child}`;
        const childMap = extractRouteMapping(options, absPath, nextPath, level + 1);
        if (Object.keys(childMap).length) {
          if (!childMap['index']) {
            childMap[options.dirkey] = nextPath;
          }
          map[options.keyTransform(child)] = childMap;
        }
      } else {
        // skip if extension doesn't match
        if (options.extensions.every((ext) => !filename.endsWith(ext))) continue;

        // transform filename to route key
        let routeKey = filename.replace(/\.[^/.]+$/, ''); // trim extension
        routeKey = options.keyTransform(routeKey); // by default camelCase-ify key

        let routeName = routeKey;
        if (routeKey === 'index') {
          // no explicit `index`
          routeName = '';
        }
        let nextPath = currentPath;
        if (routeName) {
          nextPath += `/${routeName}`;
        }

        if (!nextPath && level === 0) {
          // root level index route
          nextPath = '/';
        }

        map[routeKey] = nextPath;
      }
    }
  }

  return map;
}

export function generateJSON(options: CollectOptions) {
  const map = extractRouteMapping(options);
  const outputPath = resolve(options.outDir, 'routes.json');
  writeFileSync(
    outputPath,
    JSON.stringify(map, null, 2), // spacing level = 2
    { encoding: 'utf-8' },
  );
}

// export function generateUtils() {

// }

export function collect(options: CollectOptions) {
  const { outDir } = options;
  try {
    if (!existsSync(outDir)) {
      mkdirSync(outDir, { recursive: true });
    }
    generateJSON(options);
  } catch (error) {
    console.error(error);
    const data = readdirSync(outDir);
    if (data.length === 0) {
      rmSync(outDir, { recursive: true });
    }
  }
}
