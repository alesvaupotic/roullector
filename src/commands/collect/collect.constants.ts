import type {
  CliCollectOptions,
  CollectOptions,
} from './collect.types';

export function CamelCasify(str: string) {
  return str
    .trim()
    .replace(/([-_\s]+.)/g, function($1) {
      return $1.replace(/[-_\s]/g, '').toUpperCase();
    })
    .replace(/^(.)/, function($1) {
      return $1.toLowerCase();
    });
}

export const defaultCollectOptions: CollectOptions = {
  inDir: 'src/routes',
  extensions: ['.svelte'],
  ignorePatterns: ['^_'],
  outDir: 'src/generated/routing',
  typescript: true,
  keyTransform: CamelCasify,
  verbose: false,
  depth: Infinity,
  dirkey: '__dir',
  utils: true,
};

export const defaultCliCollectOptions: CliCollectOptions = {
  ...defaultCollectOptions,
  extensions: defaultCollectOptions.extensions.join(','),
  ignorePatterns: defaultCollectOptions.ignorePatterns.map((pattern) => (pattern instanceof RegExp) ? pattern.source : pattern).join(','),
  depth: 'Infinity',
  keyTransform: 'camelCase',
};

export const keyTransformCliToFunc: Record<CliCollectOptions['keyTransform'], (str) => string> = {
  camelCase: CamelCasify,
};
