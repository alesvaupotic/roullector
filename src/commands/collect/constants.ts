import type { CollectOptions } from './types';

export const defaultCollectOptions: CollectOptions = {
  pattern: 'src/routes',
  outDir: 'src/generated/routing',
  typescript: true,
};
