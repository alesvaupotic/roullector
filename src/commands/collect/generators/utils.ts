/* eslint-disable no-useless-escape */
import {
  writeFileSync,
  readFileSync,
} from 'fs';
import { resolve } from 'path';

import { transpile } from 'typescript';

import type { CollectOptions } from '$commands/collect/collect.types';

export function generateUtils(options: CollectOptions) {
  const { outDir, typescript } = options;
  let source = '';

  const ts = readFileSync(resolve(__dirname, './route.ts'), 'utf-8');

  if (typescript) {
    source =
`export { default as AppRoutes } from './routes.json';
${ts}
`;
  } else {
    const js = transpile(ts);
    source =
`export { default as AppRoutes } from './routes.json';
}
${js}
`;
  }

  const outputPath = resolve(outDir, 'index.' + (typescript ? 'ts' : 'js'));

  writeFileSync(
    outputPath,
    source,
    { encoding: 'utf-8' },
  );
}
