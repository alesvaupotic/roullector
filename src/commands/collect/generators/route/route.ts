import {
  writeFileSync,
  readFileSync,
  existsSync,
  mkdirSync,
} from 'fs';
import { resolve } from 'path';

import { MAPPING_FILENAME } from '$commands/collect/collect.constants';
import type { CollectOptions } from '$commands/collect/collect.types';

export function generateRouteUtil(options: CollectOptions): string|null {
  const { output, outDir, typescript } = options;

  const ext = typescript ? '.ts' : '.js';

  /**
   * get the source in relative path to build target
   * see rollup.config.ts for more details
   */
  const source = readFileSync(resolve(__dirname, '../route.source' + ext), 'utf-8');

  const data =
`export { default as AppRoutes } from './${MAPPING_FILENAME}';
${source}
`;

  let file = null;
  if (output) {
    if (!existsSync(outDir)) {
      mkdirSync(outDir, { recursive: true });
    }
    file = resolve(outDir, 'index' + ext);
    writeFileSync( file, data, { encoding: 'utf-8' });
  } else {
    file = data;
  }

  return file;
}
