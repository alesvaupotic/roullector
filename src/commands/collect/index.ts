import { Command } from 'commander';

import { defaultCollectOptions } from './constants';
import type { CollectOptions } from './types';

export function collect() {
  return new Command('collect')
    .description('collect route data')
    .option(
      '-p, --pattern <type>',
      'glob pattern to search for files',
      defaultCollectOptions.pattern
    )
    .option(
      '-o, --outDir <type>',
      'outputs generated files to this folder',
      defaultCollectOptions.outDir
    )
    .option(
      '-t, --typescript',
      'outputs generated files in typescript',
      defaultCollectOptions.typescript
    )
    .action((rawOptions: CollectOptions) => {
      const options = {
        ...defaultCollectOptions,
        ...rawOptions,
      };
    });
}
