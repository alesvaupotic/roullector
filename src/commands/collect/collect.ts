import {
  readdirSync,
  mkdirSync,
  existsSync,
  rmdirSync,
} from 'fs';

import type {
  CollectOptions,
  CollectOutput,
} from '$commands/collect/collect.types';
import { generateJSON } from '$commands/collect/generators/json';
import { generateRouteUtil } from '$commands/collect/generators/route';

import { defaultCollectOptions } from './collect.constants';

export function collect(options: CollectOptions = defaultCollectOptions): CollectOutput {
  const { outDir, output } = options;
  const result = {
    json: null,
    route: null,
  };
  try {
    if (!existsSync(outDir) && output) {
      mkdirSync(outDir, { recursive: true });
    }
    result.json = generateJSON(options);
    if (options.utils) {
      result.route = generateRouteUtil(options);
    }
  } catch (error) {
    console.error(error);
    if (existsSync(outDir) && readdirSync(outDir).length === 0) {
      rmdirSync(outDir, { recursive: true });
    }
    throw error;
  }
  return result;
}
