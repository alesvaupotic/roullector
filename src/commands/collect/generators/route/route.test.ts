import { resolve } from 'path';

import {
  fs as memfs,
  vol,
} from 'memfs';

import { defaultCollectOptions } from '$commands/collect/collect.constants';
import type { CollectOptions } from '$commands/collect/collect.types';
import { generateRouteUtil } from '$commands/collect/generators/route';
import { route } from '$commands/collect/generators/route/source';

jest.mock('fs', () => {
  const fs = jest.requireActual('fs');
  const sourcePath = './source.ts';
  const sourceOfRouteUtil = fs.readFileSync(resolve(__dirname, sourcePath), 'utf-8');
  vol.fromJSON({
    [sourcePath]: sourceOfRouteUtil,
  }, __dirname);
  return memfs;
});

describe('map path with route util', () => {
  test('static path: return same', () => {
    const path = '/a/literal/static/path';
    expect(route(path)).toBe(path);
  });

  test('dynamic path: argument is correctly injected', () => {
    const parent = '/a-path-with';
    const id = 'a-unique-id-123-124';
    expect(route(`${parent}/[arg]`, id)).toBe(`${parent}/${id}`);
  });

  test('dynamic path: multiple arguments are injected as specified', () => {
    const path = '/a-path-with/[arg_snake]/and/[arg-kebab]/and/[argCamel]';
    const argSnake = 'snake';
    const argKebab = 'kebab';
    const argCamel = 'camel';

    const routed = route(path, argSnake, argKebab, argCamel);
    expect(routed).toBe(`/a-path-with/${argSnake}/and/${argKebab}/and/${argCamel}`);
  });
});

describe('generate route utils:', () => {
  const savedJSON = vol.toJSON();
  beforeEach(() => {
    vol.fromJSON(savedJSON);
  });
  afterEach(() => {
    vol.reset();
  });

  const generateRouteUtilWithOptions = (snapshotToMatch: string, options: Partial<CollectOptions> = {}) => {
    const outputPath = generateRouteUtil({
      ...defaultCollectOptions,
      ...options,
    });
    const generated = memfs.readFileSync(outputPath, 'utf-8');
    expect(generated).toMatchSnapshot(snapshotToMatch);
  };

  test('with default options', () => {
    generateRouteUtilWithOptions('generateRouteUtil: default');
  });

  test('to javascript only', () => {
    generateRouteUtilWithOptions('mockImplementation: javascript', { typescript: false });
  });
});

