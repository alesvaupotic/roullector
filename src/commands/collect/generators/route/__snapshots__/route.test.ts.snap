// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generate route utils: to javascript only: mockImplementation: javascript 1`] = `
"export { default as AppRoutes } from './routes.json';
\\"use strict\\";
Object.defineProperty(exports, \\"__esModule\\", { value: true });
exports.route = void 0;
/**
 * build a complete path with injected arguments
 * @param path {string} based path
 * @param args {string[]} arguments to inject
 */
function route(path) {
    var _a;
    var args = [];
    for (var _i = 1; _i < arguments.length; _i++) {
        args[_i - 1] = arguments[_i];
    }
    var params = (_a = path.match(/\\\\[[a-zA-Z_-]+\\\\]/g)) !== null && _a !== void 0 ? _a : [];
    for (var i in params) {
        path = path.replace(params[i], args[i]);
    }
    return path;
}
exports.route = route;

"
`;

exports[`generate route utils: with default options: generateRouteUtil: default 1`] = `
"export { default as AppRoutes } from './routes.json';
/**
 * build a complete path with injected arguments
 * @param path {string} based path
 * @param args {string[]} arguments to inject
 */
export function route(path: string, ...args: string[]): string {
  const params = path.match(/\\\\[[a-zA-Z_-]+\\\\]/g) ?? [];
  for (const i in params) {
    path = path.replace(params[i], args[i]);
  }
  return path;
}

"
`;
